Add RFD 134: Cloud first deployments (#28072)
gravitational/teleport: rfd/0134-cloud-first-deployment.md

## What

This RFD describes the Cloud-first deployment approach, as well as the release
process changes required to support it.

## Why

Our current feature delivery process is not compatible with the company's shift
to "SaaSify" Teleport. The new (or rather, extended) Cloud-focused process aims
to address shortcomings like:

- Long release cycles where marquee new features get delivered to Cloud users
  (roughly) once a quarter and are tied to major self-hosted releases.
- Slow and monolithic release pipeline that always builds all artifacts and
  does not allow for very fast Cloud rollouts.
- Engineers being disconnected from Cloud deployments and unaware of when their
  changes are getting rolled out.

## Self-hosted release process (current)

Before diving into the new process, let's set the stage with a high-level overview
of the current release process.

Current release versioning scheme follows semver, with quarterly major releases
and monthly minor releases.

Major releases are getting rolled onto Cloud quarterly, with a month's delay.
Major features are reserved for major releases, which leads to Cloud customers
getting major features with a month's delay.

> The release process is triggered on a Github tag creation and builds the full

Reviewer: ```suggestion
The release process is triggered on a GitHub tag creation and builds the full
```

suite of artifacts that takes 2-3 hours so it's unrealistic to do multiple
deploys a day.

Cloud upgrade happens once a week to a release published the previous week, and
engineers oftentimes don't have a clear idea of when their change is being
deployed.

## Cloud release process

> In the Cloud-first world, release that's running on Cloud does not necessarily

Reviewer: ```suggestion
In the Cloud-first world, every release that's deployed on Cloud does not necessarily
```

have a corresponding self-hosted release tag. For this reason, we can't reuse
the current versioning scheme and apply it to both Cloud and self-hosted.

Instead, we want to get to a place where Cloud and self-hosted releases are
decoupled, with Cloud releases being regularly cut from the main development
branch and self-hosted releases cut from their respective release branches.

Although the concept of semantic versioning loses most of its meaning with the
continuous release train, we do need to keep it for Cloud releases to be able
to use automatic upgrades system we built.

With this in mind, the Cloud version will consist of:

> - Major: "development" version number on master branch (latest stable release + 1).

Reviewer 2: ```suggestion
- Major: "development" version number on master branch (latest stable self-hosted release + 1).
```

- Minor/patch: always 0.

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Reviewer: > might as well drop semver altogether

Our auto upgrades machinery requires valid semver, which is why we're using 14.0.0 even though the minor/patch versions don't convey anything.

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Reviewer 3: A reason for the suggestion is that semver makes little sense for CD, it's usually just a bother, plus I imagine we could patch the tooling to do what we wanted. That said we can surely live the zeroes. The argument for the commit hash is a bit stronger, I think, but up to y'all.

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Reviewer: I don't have strong feelings about the commit hash - it has the nice property of being easily tied to a commit, but it does make it harder to know at a glance if a tenant running `cloud.9da628ffcb` is newer or older than one running `cloud.27941f402`.

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Author: If we use hash, then we lose ability to compare and our auto-upgrades stop working. I think counter is fine for now, you can always lookup exact commit by tag.

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Reviewer 3: Gotcha, makes sense. Feel free to resolve the thread.

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Reviewer 4: Adding to this can of worms to suggest (perhaps, terribly) a timestamp rather than a counter, e.g. `v14.0.0-cloud.2023-07-12T04:05:48Z`

Pro:
- still easy to identify newer vs. older revisions of Teleport
- still sorts in temporal order of releases
- makes it possible to speak about compatibility windows in terms of _X days_ or _Y weeks_, which may make more sense to humans than _X releases_, especially if releases are irregular.

Con:
- I have no science to back up the idea that talking in wall-clock time is any more useful than talking in releases
- It's very verbose
- It's just a bit odd

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Reviewer 5: Cloud has been tagging infra releases like `prod-aabbccd-20230712T123633`, we could flip it around and use `v14.0.0-cloud.20230712T123633.aabbccd` to have both time-based ordering that matches the lexicographic one and a reference for the exact commit that's running.

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Reviewer 3: +1 for commit+timestamp - it has the bulk of what we care about without the need for counter bookkeeping. (Extra points if on rollbacks we don't "update" the timestamp.)

> - Prerelease: `-cloud.X` with a perpetually increasing "build" number which will

Author: Sounds good, I agree this looks like better versioning scheme. Updated.

  provide enough semver comparison guarantees for auto-upgrades to work.

Once it is time for the next self-hosted major release (every 3-4 months), the
new release branch will cut and the Cloud release version major component will
increment.

                                14.0.0     14.0.1     14.0.2
                     branch/v14 |----------|----------|---------....---->
                                |
                                |
            14.0.0-cloud.1      14.0.0-cloud.10    15.0.0-cloud.11    15.0.0-cloud.12    16.0.0-cloud.13

> master -----|-------....--------|------------------|-------....-------|------------------|--------------....---->

Reviewer 2: (Minor) Did we talk about why cloud.X version doesn't reset when we increment the major version? E.g. why `15.0.0-cloud.11` rather than `15.0.0-cloud.0`?

Author: I don't think we achieved consensus on it but I agree resetting the counter makes more sense than keep forever incrementing it. Updated.

                                                                      |
                                                                      |
                                                           branch/v15 |---------|---------|---------...---->

>                                                                       15.0.0    15.0.1    15.0.2

Reviewer 2: This doesn't render.

Reviewer: Yeah, gotta put the whole thing in ```

Once a self-hosted release branch is cut, subsequent self-hosted minor/patch
releases are published without any changes to the current process, but on a
less frequent cadence than Cloud releases off of master.

> ### Cloud release artifacts

Author: Yep, it's covered by the clients auto-upgrades section below (which is a separate work item tracked in https://github.com/gravitational/cloud/issues/4880).

The self-hosted release and promotion process is slow and takes 2-3 hours
because it builds a full suite of artifacts. Cloud users need most of these
artifacts as well for the agents they run on their own infrastructure, however
Cloud control plane needs a very limited subset of all artifacts. For this
reason, the Cloud release and promotion process will be split in 2 parts:

- The workflow that builds artifacts required for the control plane: amd64
  binaries and Docker images.
- The workflow that builds remaining artifacts such as deb/rpm packages and
  Helm charts.

> This will allow us to produce rapid releases that affect only control plane
> services (auth, proxy) while still having ability to push remaining artifacts
> when needed.

Reviewer 2: ```suggestion
This will allow us to produce rapid releases that affect only control plane
services (auth, proxy) while still having the ability to push remaining artifacts
when needed.
```

Both workflows will run in Github Actions and will reuse existing bits of the
full self-hosted release pipeline.

### Cloud promotion process

The Cloud release promotion process will be staged, with a different class of
tenants being upgraded at each stage:

1. Canary
2. Internal
3. Trials

> 4. Enterprise

Reviewer 2: ```suggestion
4. Team & Enterprise
```

The canary tenants will be a set of precreated staging clusters with some data
in them which engineers will use for connecting their agents and testing their
features. Canary clusters will be updated automatically once a Cloud release is

> cut.

Reviewer 2: I don't think we should use the term "staging" here so we don't confuse it with the cloud staging environment. I'd just drop staging.

> Before promoting the release further, engineers will test their changes and
> explicitly sign off on any significant change that's being deployed by applying
> a label (e.g. `cloud/verified`) on a master PR corresponding to this change.
> The release manager then will approve the further rollout.

Reviewer 2: ```suggestion
Before promoting the release further, engineers will test any significant change in the cloud staging environment and then explicitly sign off that the change is ready for deployment by applying a label (e.g. `cloud/verified`) on the master PR corresponding to the change.
The release manager then will approve the further rollout.
```

Once the change is rolled out to production tenants, engineers will be expected
to monitor error rates and Cloud metrics relevant to their changes.

Let's now look at some of the edge case scenarios in the release process and
how we'll be handling them.

> ### Behavior-incompatible changes

Author: Yep, added a section about it.

> ### Behavior-incompatible changes

Reviewer 3: Thanks!

Once it is time for the next self-hosted major release (every 3-4 months), the
new release branch will cut and the Cloud release version major component will
increment.

```
                                14.0.0     14.0.1     14.0.2
                     branch/v14 |----------|----------|---------....---->
                                |
                                |
            14.0.0-cloud.1      14.0.0-cloud.10    15.0.0-cloud.0     15.0.0-cloud.1     16.0.0-cloud.0
master -----|-------....--------|------------------|-------....-------|------------------|--------------....---->
                                                                      |
                                                                      |
                                                           branch/v15 |---------|---------|---------...---->
                                                                      15.0.0    15.0.1    15.0.2
```

Once a self-hosted release branch is cut, subsequent self-hosted minor/patch
releases are published without any changes to the current process, but on a
less frequent cadence than Cloud releases off of master.

> 

Reviewer 4: Will changes to the self-hosted release branches generally follow the same develop-on-master/backport-to-branch workflow we're using now?

Author: Yep, added a sentence to clarify this.

when needed.

Both workflows will run in Github Actions and will reuse existing bits of the
full self-hosted release pipeline.

### Cloud promotion process

The Cloud release promotion process will be staged, with a different class of
tenants being upgraded at each stage:

1. Canary
2. Internal
3. Trials
4. Team & Enterprise

The canary tenants will be a set of precreated clusters with some data in them
which engineers will use for connecting their agents and testing their features.
Canary clusters will be updated automatically once a Cloud release is cut.

Before promoting the release further, engineers will test any significant change
in the cloud staging environment and then explicitly sign off that the change is
ready for deployment by applying a label (e.g. `cloud/verified`) on the master
PR corresponding to the change. The release manager then will approve the
further rollout.

Once the change is rolled out to production tenants, engineers will be expected
to monitor error rates and Cloud metrics relevant to their changes.

Let's now look at some of the edge case scenarios in the release process and
how we'll be handling them.

### Behavior-incompatible changes

> Teleport's component compatibility guarantees become somewhat obsolete in a

Reviewer 4: Yes and no. If we're still cutting self-hosted releases, then people are going to expect some level of backwards compatibility between "major releases" of the self-hosted product. 

We will need to define and manage our backwards compatibility guarantees in a way that works for both the cloud-first and self-hosted release cadences (even if the self-hosted compatibility guarantee turns out to be "well, good luck with that...".)

Author: Right, this is referring to Cloud compatibility specifically. For self-hosted releases I don't expect we can change our compatibility guarantees, added a sentence to clarify this below.

Cloud-first world. Historically, we've been reserving behavior-incompatible
changes for major releases. The concept of a major release, however, becomes
non-existent in the Cloud-first world where any change, backwards compatible
or not, gets deployed to all users within days.

When applied to Cloud release model, the compatibility requirement becomes:

> "Any Cloud release must be always compatible with the previous Cloud release"

Reviewer 4: If releases are common and irregular, perhaps defining the guarantee in terms of wall-clock time (e.g. must be compatible with Teleport from _n_ days|weeks|whatever ago) may be an easier way to think of things?

> "Any Cloud release must be always compatible with the previous Cloud release"

Author: I've updated this to clarify that a release must be compatible with the oldest connected auto-upgraded agent.

Compatibility requirements affect control plane (proxy, auth and other services
we host for our users), agents and clients:

- Control plane. We have full control over it, deploy in tandem and make sure
  these are always compatible with each other.
- Agents deployed by users. We always keep backwards compatibilithy between
  adjacent releases and rely on automatic upgrades to make sure they don't drift.
- Clients. Clients require automatic upgrades to avoid breakage on behavior
  incompatible changes. This is [in progress](https://github.com/gravitational/cloud/issues/4880).

### Rollbacks

Bad releases can happen even with the multi-staged rollout process. In that
situation the process will be:

> - For the control plane, we will prefer rolling forward in most cases to a new

Reviewer 3: > we will prefer rolling forward

A word of caution, this goes against some of the more solid oncall practices I've seen elsewhere. If a rollback would solve the situation, say by avoiding a newly-added bad behavior, then it's one of the first actions I'd expect a oncall to take. It'll not be easy to get us into a state of where rollbacks are mainly safe, but it is doable and a worthy objective to have.

I don't expect changes to the design doc, as I'm sure there's not only one way of doing things, but I think this is worth saying nonetheless.

Reviewer 4: How will we manage restoring migrated resources in the event of a rollback? Do we have to look at having more sophisticated tooling for managing our resource schemas than just upgrade-on-first-run? Is this part of the reason for favouring a "roll-forward" approach?

Reviewer 3: Trunk-based CD is inherently different from our milestone-based approach. We'll push much more frequently, meaning we shouldn't hit as many migrations as folks do when bumping majors, for example.

That said, we'll have to be more careful with migrations, for sure. I won't go about listing alternatives here, lest the thread becomes and endless parade of what-ifs; instead I'll say this is a known problem with many viable solutions.

> Is this part of the reason for favouring a "roll-forward" approach?

I suppose so? It's a weak argument, I think. Leaving our customers without ways to rollback is just as bad as removing to option to ourselves.

Author: I don't think we really intend for this to be a hard prescription, "we only roll forward". It is more of a guidance I guess: if our Cloud release pipeline is quick and, say, the issue only affects the control plane, it might be cheaper/faster/less risky for us to just fix the issue, push a new release, upgrade Cloud and trigger auto-upgrades, than doing rollbacks. I think in practice we'll just decide on a case-by-case basis.

  release with the addressed issue or a rolled back change. The control plane
  release process described above will be fast enough to be able to push the
  new release quickly.
    - In a scenario where a major critical path is affected and a full "roll
      forward" process is not fast enough, the deployment team will manually
      perform release rollback using the same process that is used today.
- For the agents, the rollback mechanism is supported via automatic upgrade
  system's critical/version endpoints described in the [RFD109](https://github.com/gravitational/teleport/blob/master/rfd/0109-cloud-agent-upgrades.md).
- For the clients, which don't support automatic upgrades yet, the details are
  TBD as part of the [automatic upgrades](https://github.com/gravitational/cloud/issues/4880)
  work but we will likely need to build some automatic downgrade mechanism in
  case a client was auto-upgraded but a corresponding Github release was pulled.

### Security releases

Existing security release process is compatible with Cloud-first deployment.

> Both `-cloud.X` and regular self-hosted tags will be published from teleport-private

Reviewer 4: Will a security release block subsequent "regular" releases if there is a disclosure/update window we need to honour before pushing the security fixes back to OSS teleport? This isn't a big deal now as releases are relatively infrequent, but may become an issue if we want to release daily/several times a week.

Reviewer 4: Incidentally - is `teleport-private` expected to remain an identical-down-to-the-hash fork of OSS `teleport`?

Is it worth defining how we intend that relationship to work now that there is scope for `teleport` and `teleport-private` to diverge more quickly, thanks to more frequent releases and larger dev team? This RFD may not necessarily be the place for that, though.

Author: No changes are expected in how we handle security releases, or the private repo. We can still upgrade Cloud control plane (and users' agents, through auto-upgrades) however often we want, we would just delay porting the fixes to OSS in case of embargo like we're doing it now.

Author: Yes, that's how the current process works as well. While the embargo is in effect, if another release needs to be published it will have to be published off of teleport-private. In practice, we would usually freeze releasing until embargo lifts unless there's an urgent fix needs to be deployed.

Reviewer 6: I guess I don't have a clear idea of two timeframes:
* How long we typically need to embargo a security release for
* How frequently we want to do cloud releases in this new model ("with Cloud releases being regularly cut from the main development branch").

If the former is somewhat longer than the latter, then security releases are going to impact the cloud release process described here which would be at odds with "Existing security release process is compatible with Cloud-first deployment."

Author: The embargo period is typically 1 week. We usually freeze public releases for that week but publish off of teleport-private if we need to (e.g. urgent fix). We will just do the same with Cloud-first.

deploys a day.

Cloud upgrade happens once a week to a release published the previous week, and
engineers oftentimes don't have a clear idea of when their change is being
deployed.

## Cloud release process

In the Cloud-first world, every release that's deployed on Cloud does not necessarily
have a corresponding self-hosted release tag. For this reason, we can't reuse
the current versioning scheme and apply it to both Cloud and self-hosted.

Instead, we want to get to a place where Cloud and self-hosted releases are
decoupled, with Cloud releases being regularly cut from the main development
branch and self-hosted releases cut from their respective release branches.

Although the concept of semantic versioning loses most of its meaning with the
continuous release train, we do need to keep it for Cloud releases to be able
to use automatic upgrades system we built.

With this in mind, the Cloud version will consist of:

- Major: "development" version number on master branch (latest stable self-hosted release + 1).
- Minor/patch: always 0.
- Prerelease: `-cloud.X.Y` where `X` will represent a timestamp of when the tag

>   was cut and `Y` the corresponding commit hash: `v14.0.0-cloud.20230712T123633.aabbcc`.

Reviewer 6: Sorry I missed the above discussion about this. Please ignore.

Once it is time for the next self-hosted major release (every 3-4 months), the
new release branch will cut and the Cloud release version major component will
increment.

```
                                                   14.0.0     14.0.1     14.0.2
                                        branch/v14 |----------|----------|---------....---->
                                                   |
                                                   |

>             14.0.0-cloud.20230712T123633.aabbcc      14.0.0-cloud...  15.0.0-cloud...    15.0.0-cloud...    16.0.0-cloud...

Author: I think it's correct. The next release after branch/v14 has been cut will be 15.0.0-xxx and so on. I fixed the indentation here a bit though as it was confusing

>             14.0.0-cloud.20230712T123633.aabbcc      14.0.0-cloud...  15.0.0-cloud...    15.0.0-cloud...    16.0.0-cloud...

Reviewer 6: On the timeline there is a 15-0-0-cloud-* release after branch/v15 has been cut. Shouldn't that second v15 cloud release be a v16 cloud release?

>             14.0.0-cloud.20230712T123633.aabbcc      14.0.0-cloud...  15.0.0-cloud...    15.0.0-cloud...    16.0.0-cloud...

Author: Yeah you're right. I've updated the diagram to be less confusing (hopefully).

branches will follow the same "develop on master, backport to release branch"
workflow we're using now.

### Cloud release artifacts

The self-hosted release and promotion process is slow and takes 2-3 hours
because it builds a full suite of artifacts. Cloud users need most of these
artifacts as well for the agents they run on their own infrastructure, however
Cloud control plane needs a very limited subset of all artifacts. For this
reason, the Cloud release and promotion process will be split in 2 parts:

- The workflow that builds artifacts required for the control plane: amd64
  binaries and Docker images.
- The workflow that builds remaining artifacts such as deb/rpm packages and
  Helm charts.

This will allow us to produce rapid releases that affect only control plane
services (auth, proxy) while still having the ability to push remaining artifacts
when needed.

Both workflows will run in Github Actions and will reuse existing bits of the
full self-hosted release pipeline.

### Cloud promotion process

The Cloud release promotion process will be staged, with a different class of
tenants being upgraded at each stage:

1. Canary
2. Internal
3. Trials
4. Team & Enterprise

> The canary tenants will be a set of precreated clusters with some data in them

Reviewer 6: Canaries are typically a subset of real users with real workloads, not a testing environment. As described here, this does not sound like a canary candidate. Should this be renamed to "Testing" or something similar?

which engineers will use for connecting their agents and testing their features.
Canary clusters will be updated automatically once a Cloud release is cut.

> Before promoting the release further, engineers will test any significant change

Reviewer 6: What is the scope of "Internal", "Trials" and "Team & Enterprise". I assume the latter is "the rest", that "Internal" is our own use of teleport. What is "Trials" and what is the promotion process and decision gates between these stages?

Author: Trials are Teleport Team trial signups. I'm not sure there should be manual gates between promotions between these, after the changes have been verified in canary (or test) clusters. The SRE team doing the rollout should be watching error/perf metrics (along with engineers for their features).

in the cloud staging environment and then explicitly sign off that the change is
ready for deployment by applying a label (e.g. `cloud/verified`) on the master
PR corresponding to the change. The release manager then will approve the
further rollout.

> Once the change is rolled out to production tenants, engineers will be expected

Reviewer 6: If it is needed to coordinate the rollout of new monitoring to be able to effectively monitor a new feature, how should the rollout of that be coordinated with the rollout of the cloud release? I'd suggest that it should be in place before promotion from Canary to Internal and that should be one of the gates to promotion.

Author: I think it depends on the kind of monitoring. For product-related metrics (like, usage) I think it's fine to deploy it together with the feature (in practice, product metrics actually even lag behind feature implementation). For performance metrics, work is being done currently to implement those for various protocols we support so we'll build dashboards we'll be able to watch during the rollout (and setup alerting in future).

to monitor error rates and Cloud metrics relevant to their changes.

Let's now look at some of the edge case scenarios in the release process and
how we'll be handling them.

### Behavior-incompatible changes

Teleport's component compatibility guarantees become somewhat obsolete in a
Cloud-first world. Historically, we've been reserving behavior-incompatible
changes for major releases. The concept of a major release, however, becomes
non-existent in the Cloud-first world where any change, backwards compatible
or not, gets deployed to all users within days.

When applied to Cloud release model, and accounted for potential delays in
agent automatic upgrades, the compatibility requirement becomes:

| Any Cloud release must be always compatible with the oldest version of an
| agent that's connected to Cloud and enrolled in automatic upgrades.

Compatibility requirements affect control plane (proxy, auth and other services
we host for our users), agents and clients:

- Control plane. We have full control over it, deploy in tandem and make sure
  these are always compatible with each other.
- Agents deployed by users. We always keep backwards compatibilithy between

>   adjacent releases and rely on automatic upgrades to make sure they don't drift.

Reviewer 6: What are adjacent releases? Two consecutive cloud releases? If so, do we need to take into account a roll-forward release used to fix a problem just deployed and not count one of those releases? Or is adjacent releases two major releases?

Author: I meant consecutive Cloud releases, yes. Consecutive major releases will keep being compatible as well though as we need to preserve compat. guarantees for self-hosted users as well. So tbh in reality I don't think the way we handle backwards compatibility will change much, esp. once we've rolled out automatic upgrades to everyone.
